### PR TITLE
Adding semantic analysis when a context is set on a method or block

### PR DESCRIPTION
**Pull request by adri09070**
The interpreter evaluated a program that is never checked by a semantic analyser.

As a result, variables don't know their scope, and its not possible lookup for these unresolved variables in a context.

To fix that, a context now ask their new method/closure to do their semantic analysis each time it changes its method/closure

### CODE REVIEW

File: DebuggableASTInterpreter/DASTContext.class.st [hunk 1]
```diff
@@ -522,7 +522,8 @@ DASTContext >> methodOrBlock [
 { #category : #accessing }
 DASTContext >> methodOrBlock: aDASTClosure [
 
-	closure := aDASTClosure
+	closure := aDASTClosure.
+	closure doSemanticAnalysis
```
Comment (guillep) [inline DebuggableASTInterpreter/DASTContext.class.st:526]: Are you sure we want to put it in the setter? I would have expected to do it for example in the method lookup, when the AST is found or created.

**Review by guillep — CHANGES_REQUESTED** [review]

File: DebuggableASTInterpreter/DASTContext.class.st (hunk 1 above)
Comment (adri09070) [reply DebuggableASTInterpreter/DASTContext.class.st:526]: I hesitated. It would be better indeed but when do we know that the AST is found or created ?

**Comment by adri09070** [comment]
done

**Review by guillep — CHANGES_REQUESTED** [review]
This is strange, now it seems we are doing the semantic analysis in four different places. Is that really intended? Maybe this causes semantic analysis to happen many times at runtime?

**Comment by adri09070** [comment]
Well it's true and false at the same time:

- There is a semantic analysis in the initialisization of DASTClosure and it should
- There is a semantic analysis in the setter of method node (or block node) in `DASTClosure` and it should because it may not have been analyzed yet. (Personnally, I would throw this setter away. To me, one DASTClosure should represent only one method AST or one block AST and we shouldn't allow to completely change the method/block it represents. Plus, I don't think this setter is used anywhere).

- `DASTClosure>>#doSemanticAnalysis` used to be a helper to do the semantic analysis of its block/method and was used by `DASTContext>>#methodOrBlock:`, when a context sets its method/block. Those 2 should be removed (I thought I had removed them)

In short, we should do the semantic analysis in `DASTClosure>>#initializeWith:`, I will remove `DASTClosure>>#doSemanticAnalysis` and its call in `DASTContext>>#methodOrBlock:` and we should discuss whether the setter `DASTClosure>>#nodeAST:` is needed or not (if yes, we do the semantic analysis in this setter too, else we remove this setter

**Comment by adri09070** [comment]
I removed `DASTClosure>>#doSemanticAnalysis` and its call in `DASTContext>>#methodOrBlock:`

File: DebuggableASTInterpreter/DASTClosure.class.st [hunk 2]
```diff
@@ -34,19 +34,14 @@ DASTClosure >> body [
...
-	
+DASTClosure >> initializeWith: aRBNode [
+
 	sourceCode := aRBNode sourceCode.
 	nodeAST := aRBNode.
-	numArgs := nodeAST arguments size
-	
+	numArgs := nodeAST arguments size.
+
+	nodeAST doSemanticAnalysis
```
Comment (guillep) [inline DebuggableASTInterpreter/DASTClosure.class.st:44]: why not `self nodeAST: aRBNode`?
Maybe I'm missing something.

Anyways, it could be merged like this and could be fixed later.
thoughts?

Comment (adri09070) [reply DebuggableASTInterpreter/DASTClosure.class.st:44]: You're right. We can call the setter and it would solve everything
